rust/schema.py

```diff
@@ -63,50 +63,118 @@ class AstNode(Locatable):
 
 
 @qltest.skip
-class Unimplemented(AstNode):
+class Unimplemented(Element):
+    """
+    The base class for unimplemented nodes. This is used to mark nodes that are not yet extracted.
+    """
     pass
 
 
 class Declaration(AstNode):
+    """
+    The base class for declarations.
+    """
+    pass
+
+
+@qltest.skip
+class UnimplementedDeclaration(Declaration, Unimplemented):
+    """
+    A declaration that is not yet extracted.
+    """
     pass
 
 
 class Module(Declaration):
-    # TODO name
+    """
+    A module declaration. For example:
+    ```
+    mod foo;
+    ```
+    ```
+    mod bar {
+        pub fn baz() {}
+    }
+    ```
+    """
     declarations: list[Declaration] | child
 
 
-@qltest.collapse_hierarchy
 class Expr(AstNode):
+    """
+    The base class for expressions.
+    """
     pass
 
 
-@qltest.collapse_hierarchy
 class Pat(AstNode):
+    """
+    The base class for patterns.
+    """
     pass
 
 
-@qltest.skip
+@rust.doc_test_signature("() -> ()")
 class Label(AstNode):
+    """
+    A label. For example:
+    ```
+    'label: loop {
+        println!("Hello, world (once)!");
+        break 'label;
+    };
+    ```
+    """
     name: string
 
 
-@qltest.collapse_hierarchy
 class Stmt(AstNode):
+    """
+    The base class for statements.
+    """
     pass
 
 
-@qltest.collapse_hierarchy
-class TypeRef(AstNode):
-    # TODO
+@rust.doc_test_signature("() -> ()")
+class Type(AstNode, Unimplemented):
```

Comment:
I actually don't agree with the rename of `TypeRef` to `Type`: these are two different things. A type reference is the way a type can be represented in the AST, so it has its place in the `AstNode` hierarchy (and is `Locatable`). A type exists outside the AST, and doesn't really have a source location. 

Also, a `TypeRef` has its grammar rules that can be a disjoint from the ways you can combine types. In Swift, for example, `T?` is sugar for `Option<T>`: there will be a `?` constructor in the `TypeRef` hierarchy, but not in the `Type` one. In Rust, `Vec<_>` is a `TypeRef`, but it's not really a type (which will have the `_` filled in, or be an error).

Talking about Swift: at some point we took the decision to _not_ extract the whole `TypeRef` hierarchy as seen by the compiler, but just model it as a pair of an actual `Type` and a location in the source code, "forgetting" how the type is actually written down in source. I think we should do the same for rust (when we do start extracting types).

Reply:
I changed to `TypeRef`, `BinaryExpr` and `LiteralExpr`.